File: frontend/lib/api/authService.ts
import apiClient from './client';
import { TokenResponse } from '@/lib/utils/types';

export const authService = {
  login: async (username: string, password: string): Promise<TokenResponse> => {
    const response = await apiClient.post<TokenResponse>('/api/auth/login', {
      username,
      password,
    });
    return response.data;
  },

  register: async (
    firstName: string,
    lastName: string,
    username: string,
    email: string,
    password: string
  ): Promise<TokenResponse> => {
    const response = await apiClient.post<TokenResponse>('/api/auth/register', {
      firstName,
      lastName,
      username,
      email,
      password,
    });
    return response.data;
  },

  refresh: async (refreshToken: string): Promise<TokenResponse> => {
    const response = await apiClient.post<TokenResponse>('/api/auth/refresh', {
      refreshToken,
    });
    return response.data;
  },

  logout: async (): Promise<void> => {
    try {
      await apiClient.post('/api/auth/logout');
    } catch (error) {
      console.error('Logout error:', error);
    }
  },
};

File: userservice/src/main/java/com/example/userservice/application/services/AuthService.java
package com.example.userservice.application.services;

import com.example.userservice.application.dtos.LoginRequest;
import com.example.userservice.application.dtos.RefreshTokenRequest;
import com.example.userservice.application.dtos.RegisterRequest;
import com.example.userservice.application.dtos.TokenResponse;
import com.example.userservice.domain.entity.role.Role;
import com.example.userservice.domain.entity.user.User;
import com.example.userservice.infrastructure.repository.role.RoleRepository;
import com.example.userservice.infrastructure.repository.user.UserRepository;
import com.example.userservice.infrastructure.security.JwtService;
import com.example.userservice.infrastructure.security.RefreshTokenService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {
    private static final String DEFAULT_ROLE = "CUSTOMER";

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final RefreshTokenService refreshTokenService;

    @Transactional
    public TokenResponse register(RegisterRequest request) {
        if (userRepository.existsByUsername(request.getUsername())) {
            throw new IllegalArgumentException("Username already exists");
        }
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("Email already exists");
        }

        Role role = roleRepository.findByName(DEFAULT_ROLE)
                .orElseGet(() -> {
                    Role entity = Role.builder().name(DEFAULT_ROLE).build();
                    return roleRepository.save(entity);
                });

        User user = User.builder()
                .username(request.getUsername())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .email(request.getEmail())
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .role(role)
                .active(true)
                .build();

        return issueTokens(userRepository.save(user));
    }

    @Transactional
    public TokenResponse login(LoginRequest request) {
        User user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() -> new IllegalArgumentException("Invalid username or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new IllegalArgumentException("Invalid username or password");
        }

        if (!user.isActive()) {
            throw new IllegalArgumentException("User account is disabled");
        }

        return issueTokens(user);
    }

    @Transactional
    public TokenResponse refresh(RefreshTokenRequest request) {
        User user = refreshTokenService.consumeRefreshToken(request.getRefreshToken());
        return issueTokens(user);
    }

    @Transactional
    public void logout(RefreshTokenRequest request) {
        refreshTokenService.revokeRefreshToken(request.getRefreshToken());
    }

    private TokenResponse issueTokens(User user) {
        String accessToken = jwtService.generateAccessToken(user);
        String refreshToken = refreshTokenService.createRefreshToken(user);

        return TokenResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .tokenType("Bearer")
                .expiresIn(jwtService.getAccessTokenTtlSeconds())
                .build();
    }
}
